File: Square root without sqrt/sqrt(x).java
class Solution {
    public int mySqrt(int x) {
        if (x < 2) return x;
        long l = 0, r = x / 2, mid;
        while (l <= r) {
            mid = (l + r) / 2;
            if (mid * mid == x) return (int) mid;
            else if (mid * mid < x) l = (int) mid + 1;
            else r = (int) mid - 1;
        }
        return (int) r;
    }
}

File: Square root without sqrt/sqrt(x).py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x
        l, r = 1
        while l <= r:
            # Calculating mid value to reduce no of operations
            m = (l + r) // 2
            # Checking if mid * mid gives the value of x
            if m * m == x:
                return m
            # If the value of m * m is less than value of x, l value needs to start from m + 1
            elif m * m < x:
                l = m + 1
            # If the value of m * m is greater than value of x, r value needs to start from m - 1
            else:
                r = m - 1
        
